components/script/dom/resizeobserver.rs

```diff
@@ -0,0 +1,277 @@
+
+        let resize_observation = ResizeObservation::new(options.box_);
+
+        self.observation_targets
+            .borrow_mut()
+            .push((resize_observation, Dom::from_ref(target)));
+
+        // Note: noting a rendering opportunity here is necessary
+        // to make /resize-observer/iframe-same-origin.html PASS.
+        let window = window_from_node(target);
+        let pipeline_id = window.pipeline_id();
+        ScriptThread::note_rendering_opportunity(pipeline_id);
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-unobserve>
+    fn Unobserve(&self, target: &Element) {
+        self.observation_targets
+            .borrow_mut()
+            .retain_mut(|(_obs, other)| !(&**other == target));
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-disconnect>
+    fn Disconnect(&self) {
+        self.observation_targets.borrow_mut().clear();
+    }
+}
+
+/// State machine equivalent of active and skipped observations.
+#[derive(Default, JSTraceable, MallocSizeOf, PartialEq)]
+enum ObservationState {
+    #[default]
+    Done,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-activetargets-slot>
+    Active,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-skippedtargets-slot>
+    Skipped,
+}
+
+/// https://drafts.csswg.org/resize-observer/#resizeobservation
+#[derive(JSTraceable, MallocSizeOf)]
+struct ResizeObservation {
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-target>
+    /// Note: `target` is kept out of here, to avoid having to root the `ResizeObservation`.
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-observedbox>
+    observed_box: ResizeObserverBoxOptions,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-lastreportedsizes>
+    last_reported_sizes: Vec<ResizeObserverSizeImpl>,
+    /// State machine mimicking the "active" and "skipped" targets slots of the observer.
+    state: ObservationState,
+}
+
+impl ResizeObservation {
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-resizeobservation>
+    pub fn new(observed_box: ResizeObserverBoxOptions) -> ResizeObservation {
+        let size_impl = ResizeObserverSizeImpl::new(0.0, 0.0);
+        ResizeObservation {
+            observed_box,
+            last_reported_sizes: vec![size_impl],
+            state: Default::default(),
+        }
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-isactive>
+    fn is_active(&self, target: &Element) -> bool {
+        let last_reported_size = self.last_reported_sizes[0];
+        let box_size = calculate_box_size(target, &self.observed_box);
+        let width = box_size.width().to_f64_px();
+        let height = box_size.height().to_f64_px();
+        (
+            last_reported_size.inline_size(),
+            last_reported_size.block_size(),
+        ) != (width, height)
+    }
```

Comment:
I think the follow is shorter and clearer:
```suggestion
        box_size.width().to_f64_px() != last_reported_size.inline_size() ||
            box_size.height().to_f64_px() != last_reported_size.block_size()
```